### PR TITLE
feat: rewrite learn urls

### PR DESCRIPTION
**Pull request by kendallstrautman**
## "Ready for Review" checklist

- [X] The Vercel preview link has been added to this PR's description
- [X] The Asana task has been added to this PR's description
- [ ] This PR's link has been added to the "📐 Pull Request" field on the Asana task
- [ ] This Vercel preview link has been added to the "🔍 Preview" field on the Asana task
- [x] The description template has been filled in below

---

## Relevant links

- [Preview link](https://dev-portal-git-ksrewrite-tutorial-urls-hashicorp.vercel.app/waypoint/tutorials) 🔎
- [Asana task](https://app.asana.com/0/1199634971449915/1201956804128036) 🎟️

## What & Why

This PR adds a remark plugin that  rewrites urls for learn tutorial content that reference other learn tutorials or collections.

If the link references a tutorial or collection that is included in the beta-product list, this rewrites the link path to be relative to dev dot.

All other tutorial and collection links are rewritten to point externally to production learn.hashicorp.com

Once dev portal is publicly available, this plugin will be deprecated as all written internal links within the content will be updated in the content itself. This is an interim workaround while we have a divergent state during beta.

### The api route

In order to rewrite tutorial urls properly to the dev portal format, we need extra information not currently included in the URL to get the collection context. In order to access this context, we need to query the API. This PR adds a new api route that creates a tutorial map for all tutorials in the API to ease this look-up. 

## How

See comments for annotation on approach. 

## Testing

- Caching: Visit the [`api/tutorials-map` vercel log page](https://vercel.com/hashicorp/dev-portal/6bscg864YkrEtxwyvwwPMJZaYVKe/functions?name=api%2Ftutorials-map). Then [go to a tutorial](https://dev-portal-i4vqlqr0s-hashicorp.vercel.app/waypoint/tutorials/get-started-kubernetes/get-started-intro) and see the logs. Make sure that the _get all tutorials_ call only happens once. 
- Link rewriting: Visit a tutorial and click on links within the content: 
- Check for those linking to waypoint or vault that stay within dev portal. 
- Check for those linking to other product tutorials or collections linking externally to learn. 
- Check for external links to be untouched. 

## Anything else?

This plugin doesn't account for docs links. We will need to address that in [follow up work. ](https://app.asana.com/0/1201972623923699/1202144257024022)

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -10,4 +10,5 @@ NEXT_PUBLIC_ALGOLIA_SEARCH_ONLY_API_KEY=5e4adfd8094367056501547d6fedb6c5
 # NEXT_PUBLIC_ALGOLIA_INDEX=product_WAYPOINT
 
 MKTG_CONTENT_API="https://mktg-content-api.vercel.app"
-# Note: check .env.local.example for additional required env vars
\ No newline at end of file
+# Note: check .env.local.example for additional required env vars
+NEXT_PUBLIC_LEARN_API_BASE_URL="https://pidr8rzmwp.us-east-1.awsapprunner.com"
```
Comment (kendallstrautman) [inline .env:14]: Officially adding this public env in the repo.

File: .jest/setup.js [hunk 2]
```diff
@@ -1 +1,9 @@
 import '@testing-library/jest-dom/extend-expect'
+import path from 'path'
+import { unflatten } from 'flat'
+import { getHashiConfig } from '../config'
+
+const env = process.env.HASHI_ENV || 'development'
+const envConfigPath = path.join(process.cwd(), 'config', `${env}.json`)
+
+global.__config = unflatten(getHashiConfig(envConfigPath))
```
Comment (kendallstrautman) [inline .jest/setup.js:9]: I need to access the beta product data from config. Lmk if this is the right approach for jest set-up.

File: src/lib/learn-client/api/tutorial/index.ts [hunk 3]
```diff
@@ -60,7 +60,10 @@ export async function getTutorials(
 // if a limit is not passed, all tutorials are fetched
 export async function getAllTutorials(
   options?: getAllTutorialsOptions
-): Promise<Tutorial[] | Pick<Tutorial, 'slug'>[]> {
+): Promise<
+  Tutorial[] | (Pick<Tutorial, 'slug'> & { collection_slug: string }[])
+> {
+  console.log('CALLING GET ALL TUTORIALS!!!!!!!!!!!!!!!!!') // testing the caching efficacy
```
Comment (kendallstrautman) [inline src/lib/learn-client/api/tutorial/index.ts:66]: TODO: remove, once approach is validated

File: src/lib/remark-plugins/rewrite-tutorial-links/utils/get-tutorial-map.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import fetch from 'node-fetch'
+
+export async function getTutorialMap() {
+  let result = {}
+  const baseUrl = process.env.VERCEL_URL
```
Comment (kendallstrautman) [inline src/lib/remark-plugins/rewrite-tutorial-links/utils/get-tutorial-map.ts:5]: Is there a different / better way to get this base url?  In looking at the docs, `VERCEL_URL` seemed like the best fit.

File: src/pages/api/tutorials-map.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  const mapData = await cachedGenerateTutorialMap()
+  res.status(200).json(mapData)
+}
+
+export async function generateTutorialMap() {
+  const allTutorials = await getAllTutorials({
+    fullContent: false,
+    slugsOnly: true,
+  })
+  console.log(
```
Comment (kendallstrautman) [inline src/pages/api/tutorials-map.ts:19]: TODO: remove once approach is validated.

File: src/pages/api/tutorials-map.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import { NextApiRequest, NextApiResponse } from 'next'
+import { getTutorialSlug } from 'views/collection-view/helpers'
+import { getAllTutorials } from 'lib/learn-client/api/tutorial'
+import moize, { Options } from 'moize'
+
+export default async function tutorialsMapHandler(
+  req: NextApiRequest,
+  res: NextApiResponse
+) {
+  const mapData = await cachedGenerateTutorialMap()
```
Comment (kendallstrautman) [inline src/pages/api/tutorials-map.ts:10]: The intent here is that we only want the `getAllTutorials` call to hit once (it's quite large), and then cache that map for subsequent requests. 

I first tried to memoize the function within the plugin definition file itself. But with ISR, it didn't seem to persist there properly. Memoizing via this API route is, to my testing knowledge, behaving as expected. For a single deploy, the `getAllTutorials` function is called on the first tutorial page hit, and then the cached map is served on subsequent hits. Check the vercel function logs while digging in the preview to validate.

File: src/pages/api/tutorials-map.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  res.status(200).json(mapData)
```
Comment (kendallstrautman) [inline src/pages/api/tutorials-map.ts:11]: I can make this response behavior more robust, I just want to validate this approach first.

File: .jest/setup.js (hunk 2 above)
Comment (BRKalow) [reply .jest/setup.js:9]: Sweetness, nice!

File: src/pages/api/tutorials-map.ts (hunk 6 above)
Comment (kendallstrautman) [reply src/pages/api/tutorials-map.ts:10]: Screen Shot 2022-04-18 at 2 30 24 PM

Comment (BRKalow) [reply src/pages/api/tutorials-map.ts:10]: These run on AWS lambdas, so there's not necessarily a guarantee that we'll get the same lambda instance on subsequent invocations. If that does happen, the in-memory caching wouldn't do anything for us.

We could set some caching headers to have this API route actually be cached by Vercel ([ref](https://vercel.com/docs/concepts/functions/edge-caching))

Code example: https://github.com/hashicorp/mktg-content-workflows/blob/main/api/assets.ts#L80

File: src/lib/remark-plugins/rewrite-tutorial-links/__tests__/rewrite-tutorial-links.test.ts [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    '/waypoint/tutorials/getting-started/getting-started-config',
+  'waypoint/get-started-ui':
+    '/waypoint/tutorials/getting-started/getting-started-ui',
+  'vault/consul-deploy': '/vault/tutorials/consul-integration/consul-deploy',
+  'waypoint/get-started': '/waypoint/tutorials/get-started-docker/get-started',
+}
+
+// TESTS -----------------------------------------------------------------
+
+describe('rewriteTutorialLinks remark plugin', () => {
```
Comment (BRKalow) [inline src/lib/remark-plugins/rewrite-tutorial-links/__tests__/rewrite-tutorial-links.test.ts:61]: 👏

File: src/lib/remark-plugins/rewrite-tutorial-links/__tests__/rewrite-tutorial-links.test.ts [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+  beforeEach(async () => {
+    const scope = nock('http://localhost:3000/api/tutorials-map')
```
Comment (kendallstrautman) [inline src/lib/remark-plugins/rewrite-tutorial-links/__tests__/rewrite-tutorial-links.test.ts:63]: TODO: update this url so the tests pass in the CI

File: src/lib/remark-plugins/rewrite-tutorial-links/index.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+//import moize, { Options } from 'moize'
+import { ProductOption } from 'lib/learn-client/types'
+import {
+  getTutorialMap,
+  handleCollectionLink,
+  handleTutorialLink,
+} from './utils'
+
+const learnProducts = new RegExp(Object.keys(ProductOption).join('|'), 'g')
+const learnLink = new RegExp('(learn.hashicorp.com|collections|tutorials)')
```
Comment (BRKalow) [inline src/lib/remark-plugins/rewrite-tutorial-links/index.ts:29]: Could we make this a bit more restrictive? Maybe test on `/collections/` and `/tutorials/` to validate those pieces are part of the path?

File: src/lib/remark-plugins/rewrite-tutorial-links/index.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+function handleRewriteTutorialsLink(node: Link | Definition) {
+  try {
+    // return early if non tutorial or collection link
+    if (!learnLink.test(node.url)) {
+      return
+    }
+
+    const [product] = node.url.match(learnProducts)
+    const isExternalLearnLink = node.url.includes('learn.hashicorp.com')
+    const isBetaProduct = __config.dev_dot.beta_product_slugs.includes(product)
```
Comment (BRKalow) [inline src/lib/remark-plugins/rewrite-tutorial-links/index.ts:50]: nit, consider using `getIsBetaProduct` from `src/lib/get-is-beta-product.ts`

**Review by BRKalow — COMMENTED** [review]
Looking good. Nice idea to use an API route for this.

File: src/pages/api/tutorials-map.ts (hunk 6 above)
Comment (kendallstrautman) [reply src/pages/api/tutorials-map.ts:10]: Oh great! Okay I'll check this out

File: src/lib/remark-plugins/rewrite-tutorial-links/index.ts (hunk 10 above)
Comment (kendallstrautman) [reply src/lib/remark-plugins/rewrite-tutorial-links/index.ts:29]: Good call, i'll tighten this up.

File: src/lib/remark-plugins/rewrite-tutorial-links/index.ts (hunk 11 above)
Comment (kendallstrautman) [reply src/lib/remark-plugins/rewrite-tutorial-links/index.ts:50]: Wonderful, will do

File: src/lib/remark-plugins/rewrite-tutorial-links/__tests__/rewrite-tutorial-links.test.ts [hunk 12]
```diff
@@ -0,0 +1,212 @@
+import nock from 'nock'
+import remark from 'remark'
+import { rewriteTutorialLinksPlugin } from 'lib/remark-plugins/rewrite-tutorial-links'
+import { expect } from '@playwright/test'
```
Comment (BRKalow) [inline src/lib/remark-plugins/rewrite-tutorial-links/__tests__/rewrite-tutorial-links.test.ts:4]: I don't think we need to use `export` from `playwright`. The global available from jest should be what we want, right?

File: src/lib/remark-plugins/rewrite-tutorial-links/utils/get-tutorial-map.ts (hunk 4 above)
Comment (BRKalow) [reply src/lib/remark-plugins/rewrite-tutorial-links/utils/get-tutorial-map.ts:5]: Does a path-only request work?

```ts
await fetch('/api/tutorials-map')
```

EDIT: Oh probably not as this runs during build too, right?

File: src/pages/api/tutorials-map.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+  res.setHeader('cache-control', `s-maxage=${MAP_MAX_AGE_IN_SECONDS}`)
+  if (Object.keys(mapData).length > 0) {
+    res.status(StatusCodes.OK).json(mapData)
+  } else {
+    res.status(StatusCodes.BAD_REQUEST).json('Failed to fetch all tutorials')
+  }
+}
+
+export async function generateTutorialMap() {
+  console.log('Generating tutorial map') // For testing caching
```
Comment (BRKalow) [inline src/pages/api/tutorials-map.ts:24]: 🗒️  to remove before merge

File: src/pages/api/tutorials-map.ts [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+import moize, { Options } from 'moize'
+
+// 1 hour
+const MAP_MAX_AGE_IN_SECONDS = 60 * 60 * 60
+
+export default async function tutorialsMapHandler(
+  req: NextApiRequest,
+  res: NextApiResponse
+) {
+  const mapData = await cachedGenerateTutorialMap()
```
Comment (BRKalow) [inline src/pages/api/tutorials-map.ts:14]: Maybe a try/catch here to handle an error case and return a 500?

File: src/pages/api/tutorials-map.ts [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    const oldPath = t.slug
+    const newPath = getTutorialSlug(t.slug, t.collection_slug)
+    return [oldPath, newPath]
+  })
+
+  return Object.fromEntries(mapItems)
+}
+
+const moizeOpts: Options = { isPromise: true, maxSize: Infinity }
+const cachedGenerateTutorialMap = moize(generateTutorialMap, moizeOpts)
```
Comment (BRKalow) [inline src/pages/api/tutorials-map.ts:40]: Do we still need/want the memoization wrapper with the endpoint caching in place?

**Review by BRKalow — COMMENTED** [review]
Thinking out loud, what's the expected behavior for the pages rendered during build? Technically the _latest_ api route would not be available at that point. The recommendation in that case might be to use the same helper during build which is used to power the API route and memoize the result.

File: src/lib/remark-plugins/rewrite-tutorial-links/__tests__/rewrite-tutorial-links.test.ts (hunk 12 above)
Comment (kendallstrautman) [reply src/lib/remark-plugins/rewrite-tutorial-links/__tests__/rewrite-tutorial-links.test.ts:4]: Oops this must have been an auto-import! I'll remove

File: src/pages/api/tutorials-map.ts (hunk 15 above)
Comment (kendallstrautman) [reply src/pages/api/tutorials-map.ts:40]: 🤔

File: src/lib/learn-client/api/tutorial/index.ts [hunk 16]
```diff
@@ -60,11 +60,13 @@ export async function getTutorials(
...
   options?: getAllTutorialsOptions
-): Promise<Tutorial[] | Pick<Tutorial, 'slug'>[]> {
+): Promise<
+  Tutorial[] | (Pick<Tutorial, 'slug'> & { collection_slug: string }[])
+> {
   let result = []
   const limit = options?.limit?.toString()
   const recurse = Boolean(!limit)
-
+  console.log('CALLING GET ALL TUTORIALS!!!!!!!!!!!!!!!!!!!')
```
Comment (kendallstrautman) [inline src/lib/learn-client/api/tutorial/index.ts:69]: TODO remove before merging

File: src/lib/remark-plugins/rewrite-tutorial-links/utils/get-tutorial-map.ts (hunk 4 above)
Comment (kendallstrautman) [reply src/lib/remark-plugins/rewrite-tutorial-links/utils/get-tutorial-map.ts:8]: It didn't work for me, I was getting an absolute url error 🤷 when trying to run this locally.

File: src/lib/remark-plugins/rewrite-tutorial-links/utils/get-tutorial-map.ts [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    : 'http://localhost:3000'
+  const apiRoute = new URL('api/tutorials-map', baseUrl)
+
+  try {
+    if (isDuringBuild) {
+      console.log('DURING BUILD -------------')
+      const tutorialMapRes = await cachedGenerateTutorialMap()
+      result = tutorialMapRes
+    } else {
+      console.log('DURING ISR ------------------')
```
Comment (kendallstrautman) [inline src/lib/remark-plugins/rewrite-tutorial-links/utils/get-tutorial-map.ts:19]: TODO remove before merging

**Comment by kendallstrautman** [comment]
I'll remove the final console.logs before moving...leaving those in for now in case anyone wants to check the logs to validate.

File: src/lib/remark-plugins/rewrite-tutorial-links/utils/get-tutorial-map.ts [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+  try {
+    if (isDuringBuild) {
+      const tutorialMapRes = await cachedGenerateTutorialMap()
+      result = tutorialMapRes
+    } else {
+      const tutorialMapRes = await fetch(apiRoute)
+      if (tutorialMapRes.ok) {
+        result = await tutorialMapRes.json()
+      }
+    }
```
Comment (BRKalow) [inline src/lib/remark-plugins/rewrite-tutorial-links/utils/get-tutorial-map.ts:28]: 👏

**Review by BRKalow — APPROVED** [review]
Looks great, nicely done here. Tricky problem!

**Comment by kendallstrautman** [comment]
Post-merging loom URLs for context - https://www.loom.com/share/f9b2e5156cb04daa9cd4163bb24b859e & https://loom.com/share/7a1011daf2a74497bd9b28cb57f5fc99

**Pull request merged**